compose_box: Implement new Send button column.

karlstolley opened the pull request:
This PR adds the styles and logic for [the redesigned Send button column](https://zulip-compose-decomposed.vladkorobov.repl.co/), which includes moving Drafts adjacent the compose box and above the send button, and expanding the tooltip content presented from the Send-adjacent vdots menu.

As of 2023-11-06, the PR includes the structural, visual, and responsive components, as well as new condensed tooltip logic.

There's one Puppeteer test that is failing more or less non-deterministically. Something is amiss with the `fill_form` or `clear_and_type` functions as called from within `edit_stream_message_draft()`. What's weird is that this failure starts to show up in the commit that's purely CSS changes that establish the basic geography of the compose box, icon bars, and new send/vdots/Draft column.

Fixes: #27576

**TODOs, including feedback from @alya:**

- [x] Place character count above send buttons
- [x] Popover delay when changing Enter-to-send preferences
- [x] Remove `<hr>` between scheduled-message items
- [x] Only show the Drafts popover item when the Drafts button is not displayed on the compose box
- [x] Set tooltip to “Compose options” -> testing "Send options," since the popover's contents are arguably more about _sending_ than _composing_
- [x] Restore purple color to send button as currently on `main`
- [ ] Fix failing Puppeteer test on drafts
- [ ] Present a better selection of screenshots with the PR
- [ ] Pending discussion: shore up visual look of button colors, shadow, and placement with "underlay" effect
**Screenshots**
![responsive-send-area-2023-11-03](https://github.com/zulip/zulip/assets/170719/5365e14d-16c1-4515-8cc3-4512c46729d8)

alya commented:
Cool! Looks like the full-screen version needs to be fixed:
<img width="805" alt="Screenshot 2023-11-03 at 5 36 37 PM" src="https://github.com/zulip/zulip/assets/2090066/2fd0be97-8f01-4d76-8017-8e5a69bc1106">

alya commented:
Also, I think if you stretch the compose box, Drafts should stay near the "Send" button? I haven't checked against @terpimost 's designs.

<img width="823" alt="Screenshot 2023-11-03 at 5 36 23 PM" src="https://github.com/zulip/zulip/assets/2090066/d2da47b7-a83b-4f5e-b6d7-cc7ff88222e4">

karlstolley commented:
Thanks @alya. What browser are you looking at it in? Both Chrome and Firefox on MacOS are correctly stretching the textarea when the compose box is made full screen (and any other sizes).

And yes, @terpimost's design calls for Drafts to follow the top of the text area. I actually grew to like it, but I've been staring at it a lot.

alya commented:
Huh, I'm using Chrome on a Mac.

alya commented:
Oh, I think it's preserving the stretching I did previously, even after I close and reopen the compose box. Try manually stretching and then expanding to see if you can replicate?

karlstolley commented:
@alya said: 

> Oh, I think it's preserving the stretching I did previously, even after I close and reopen the compose box. Try manually stretching and then maximizing the compose box to see if you can replicate?

Yup. That was the issue--it should be corrected now in a new set of commits, but let's keep an eye on it.

alya commented:
Thoughts from with the new three-dot message tooltip:

1. Let's drop the line between the scheduled message options, since they are now in the same category.
2. Switching between the two options for how to send the message feels confusing -- you aren't sure if it worked or not. Probably the simplest improvement is to add some delay before closing the menu, and see if still feels weird.
3. Having "Drafts" in the menu seems unnecessary to me as long as the window is wide enough to show it directly in the compose box.

karlstolley commented:
@alya said:

> Thoughts from with the new three-dot message tooltip:
> 
>     1. Let's drop the line between the scheduled message options, since they are now in the same category.
> 
>     2. Switching between the two options for how to send the message feels confusing -- you aren't sure if it worked or not. Probably the simplest improvement is to add some delay before closing the menu, and see if still feels weird.
> 
>     3. Having "Drafts" in the menu seems unnecessary to me as long as the window is wide enough to show it directly in the compose box. In super narrow windows, I'd probably put "Drafts" last.

100% with you on all of those. My thinking is that those each probably deserve more focused follow-up PRs, since this one is already so big--although I can see how much of a pain it might be to introduce a delay onto the enter-to-send options, as that's both pretty important and also hopefully pretty minor.

alya commented:
3. The tooltip needs to be changed; something like "Compose options" might do.
4. The selection box on the menu is too wide on the left.
5. This is not a regression (so could be a follow-up), but let's immediately select the first option when opening the menu with the keyboard, just like the `i` keyboard shortcut does. (We also had a similar discussion for the gear menu, etc. recently.)
![Screenshot 2023-11-06 at 1 20 02 PM](https://github.com/zulip/zulip/assets/2090066/4e4ac4ec-1169-4bc2-b4bd-c3854a9dca16)

karlstolley commented:
Thanks @alya! Can you explain what you mean here?

> 4. The selection box on the menu is too wide on the left.

alya commented:
The selection box in the screenshot above is visually hidden under the "Send" button, and also is not centered around the `...` icon.

alya commented:
I don't think we should change the color of the button right now, or if we do, we ought to change it in more places. Currently, the color matches our popover menu confirmation buttons.

<img width="251" alt="Screenshot 2023-11-06 at 10 31 01 PM" src="https://github.com/zulip/zulip/assets/2090066/906c0215-8a23-4afc-865e-ef55bd1a7b0a">

karlstolley commented:
Hi @alya! I've addressed your feedback as listed in the TODOs at the top of the PR. Maybe you want to have another look, and we can take the discussion of the remaining things to CZO tomorrow?

alya commented:
Thanks! Can we put "Drafts" last when it's in the three-dot menu (on narrow screens)? It feels like a more natural place to add an option.

karlstolley commented:
> Thanks! Can we put "Drafts" last when it's in the three-dot menu (on narrow screens)? It feels like a more natural place to add an option.

Yes! The commit with that is up, along with a fresh push to this PR in general. Still have that lingering failing test, and I need to open up a CZO design discussion about the button layout and shadow effects.

timabbott commented:
Hmm, the dark theme Drafts in narrow windows feels a bit adrift here (not clearly horizontally aligned with anything):

![image](https://github.com/zulip/zulip/assets/2746074/62ae7482-bbd0-4fbe-a11d-b316749e1e5e)

timabbott commented:
Test-deploying this on chat.zulip.org.

karlstolley commented:
> Hmm, the dark theme Drafts in narrow windows feels a bit adrift here (not clearly horizontally aligned with anything):
> 
> 
> 
> ![image](https://github.com/zulip/zulip/assets/2746074/62ae7482-bbd0-4fbe-a11d-b316749e1e5e)
> 
> 

It's like that in light theme too, but both are faithful to Vlad's design. The alignment is only apparent when it's hovered or keyboard-focused and then the outline of a button area becomes apparent. But I agree this doesn't look quite right

timabbott commented on web/templates/compose.hbs:
```diff
@@ -90,9 +90,9 @@
-                                        <span>{{t 'Send' }}</span>
+                                        <i class="zulip-icon zulip-icon-send"></i>
                                     </button>
                                     <button class="animated-purple-button message-control-button" id="send_later" tabindex=0 type="button" data-tippy-content="{{t 'Send later' }}">
```
This should be a follow-up, but the `message-control-button` class is probably a bad name It's a name overlap with the "message controls buttons" that hover in the message feed in `web/templates/message_controls.hbs`. I think `send-control-button` would be better given how this is used?

karlstolley replied:
Yes, `send-control-button` makes a lot more sense to me. And that can probably be cleaned up at the same time as the `compose_control_button` class on the Drafts button, as that's hanging around only for the sake of its Tippy tooltip.

timabbott commented on web/src/scheduled_messages_popover.js:
```diff
@@ -17,6 +17,7 @@ import {parse_html} from "./ui_util";
 import {user_settings} from "./user_settings";
 
 export const SCHEDULING_MODAL_UPDATE_INTERVAL_IN_MILLISECONDS = 60 * 1000;
+const ENTER_SENDS_SELECTION_DELAY = 600;
```
@terpimost @alya FYI this 600ms delay when you click a different option in the "Enter sends" popover is sorta a unique value in the app. I don't have a clear idea in my mind what this would want to be consistent with or how we want feedback for selecting that option in this popover to work, exactly (In a lot of other places, we just close the popover on click), but I think we can settle this as a follow-up.

timabbott commented:
Merged, thanks @karlstolley and @terpimost!

terpimost commented:
Sorry for the slow response. I support the merge and think that details of resposivity and other polishments could be done and should be reviewed once we move actual design.
In my proposal minimal breakpoint before going to mobile view is this:
<img width="482" alt="image" src="https://github.com/zulip/zulip/assets/1903309/92acb9cb-c895-45d7-a11d-5846b410203e">
<img width="495" alt="image" src="https://github.com/zulip/zulip/assets/1903309/835572f0-f9f4-4381-a052-2bad18fba8af">
Note how Button label is aligned and how hover area is aligned. 
The next breakpoint hides Drafts under more button:
<img width="625" alt="image" src="https://github.com/zulip/zulip/assets/1903309/77a5e8ce-e6e8-4b53-acaa-73654df76fa4">
<img width="239" alt="image" src="https://github.com/zulip/zulip/assets/1903309/339569b4-c695-48a2-928d-f8c664c52e5e">

One thing we should "updesign" later is to add a little delay and transition for loader... probably the loader design itself. Right now every time user clicks there is a blink of loader for a fraction of a second
